src/lib/HST/Frontend/Transformer/Messages.hs

```diff
@@ -24,3 +26,142 @@ skipNotSupported
+--   code specified by the given source span.
+skipNotSupportedWithExcerpt
+  :: Members '[InputFile, Report] r
+  => String      -- ^ The name of the feature (plural) that is not supported.
+  -> S.SrcSpan a -- ^ The source span of the excerpt to display.
+  -> Sem r ()
+skipNotSupportedWithExcerpt feature srcSpan = do
+  excerpt <- displayCodeExcerpt srcSpan
+  report
+    $ Message Info
+    $ feature ++ " are not supported and will be skipped!\n" ++ excerpt
+
+-- | Displays an excerpt of the input code specified by the given source span.
+--
+--   The excerpt consists of an introductory line with the file path and the
+--   source span numbers, all lines of the input program that are at least
+--   partially contained in the given source span, including their line
+--   numbers, and marks showing the start and end of the spanned code.
+displayCodeExcerpt :: Member InputFile r => S.SrcSpan a -> Sem r String
+displayCodeExcerpt S.NoSrcSpan = return "No source span information available!"
+displayCodeExcerpt src@S.SrcSpan {} = do
+  maybeContent <- getInputFile (S.srcSpanFilePath src)
+  return $ case maybeContent of
+    Nothing      ->
+      "No content was found for the file `" ++ S.srcSpanFilePath src ++ "`!"
+    Just content ->
+      let ls = getLines (S.srcSpanStartLine src - 1) (S.srcSpanEndLine src)
+            content
+      in if not (isValidSrcSpan ls)
+           then "The source span "
+             ++ srcString
+             ++ " of `"
+             ++ S.srcSpanFilePath src
+             ++ "` cannot be fully displayed!"
+           else S.srcSpanFilePath src
+             ++ ':' : srcString ++ ":\n" ++ unlines (prettyLines ls)
+ where
+  -- | Returns a sublist of the given list specified by the given indices.
+  --
+  --   The first of the zero-based indices is inclusive, the second is
+  --   exclusive. Invalid indices do not cause runtime errors.
+  --   Example: getLines 1 3 [1, 2, 3, 4, 5] = [2, 3]
+  getLines :: Int -> Int -> [a] -> [a]
+  getLines i1 i2 = take (i2 - i1) . drop i1
+
+  -- | Tests if a source span is valid by checking if the locations specified
+  --   in the source span do exist in the given lines.
+  isValidSrcSpan :: [String] -> Bool
+  isValidSrcSpan [] = False
+  isValidSrcSpan ls
+    = (length ls == S.srcSpanEndLine src - S.srcSpanStartLine src + 1)
+    && (length (head ls) >= S.srcSpanStartColumn src)
+    && (length (last ls) >= S.srcSpanEndColumn src - 1)
+
+  -- | Builds a string displaying the line and column of the source span start
+  --   and end.
+  srcString :: String
+  srcString = show (S.srcSpanStartLine src)
+    ++ ':'
+    : show (S.srcSpanStartColumn src)
+    ++ '-' : show (S.srcSpanEndLine src) ++ ':' : show (S.srcSpanEndColumn src)
+
+  -- | Adds line numbers to each given line and adds marks showing the start
+  --   and end of the spanned code.
+  prettyLines :: [String] -> [String]
+  prettyLines ls
```

Comment:
I'm not sure how helpful it is to show all lines of multi-line source spans. The GHC deals with multi-line source spans by showing only the first line. The `^^^` markers are followed by an ellipsis in this case to indicate that the source span has been truncated. 

```
Error: Enumerations are not supported!
head.hs:8:8-10:16:
   |
 8 | test = [1
   |        ^^...
```

Of course this is not perfect as the relevant information may be hiding in the truncated lines. However, when all lines are shown, the actual error message gets obfuscated. Since the text of the error message is more important than the code excerpt (which the user can simply lookup in their editor), I'd recommend to adopt the GHC's approach.

Reply:
This is not a big problem currently, because these excerpts are currently only shown for fatal errors, so we have a maximum of one excerpt per program call. I've written something about this in the pull request description, so please read it.
And while the GHC does display only one regular line of the input code, it practically displays much more in many cases describing the location of the error (I have constructed an example where it prints out as much as I want). I definitely think that we should display more than one line, but we could set a maximum in some way. One idea I have is to display up to five lines and to display the first two and last two lines of longer excerpts, like this:
```
head.hs:8:8-101:11:
             vvvvv
  8 | test = [1
  9 |           ..
  ⁝
100 |    5
101 |          ] ++ [6]
         ^^^^^^^
```

Reply:
> This is not a big problem currently, because these excerpts are currently only shown for fatal errors

If multi-line code excerpts prevent us from having code excerpts in non-fatal messages, we do have a big problem :D Especially in case of non-fatal messages the user needs enough information to quickly decide whether they want to investigate the warning but not too much information that they miss the warning altogether.

> the GHC [...] practically displays much more in many cases describing the location of the error

While I am aware that the GHC often prints additional code (using it's pretty printer) in messages like "In a stmt of a 'do' block: …", the actual code excerpts (i.e., ["caret diagnostics"](https://hackage.haskell.org/package/ghc-8.10.1/docs/src/ErrUtils.html#getCaretDiagnostic)) include the first line only as far as I know.

> One idea I have is to display up to five lines and to display the first two and last two lines of longer excerpts

That would be a viable option. Alternatively, we could show the first line by default and add a command line flag to show multi-line excerpts.

Reply:
> If multi-line code excerpts prevent us from having code excerpts in non-fatal messages, we do have a big problem :D

Multi-line excerpts are not the reason I didn't implement code excerpts for `skipNotSupported` yet. I think we should offer an option to disable these warnings altogether.